scanner/src/test/kotlin/storages/ClearlyDefinedStorageTest.kt

```diff
@@ -0,0 +1,393 @@
+            assertValidResult(storage.read(pkg, SCANNER_DETAILS))
+        }
+
+        "use information from a source artifact if available" {
+            val sourceArtifact = RemoteArtifact("https://source-artifact.org/test", Hash.NONE)
+            val expUrl = PACKAGE_URL.replace("maven/", ComponentType.SOURCE_ARCHIVE.value + "/")
+            val pkg = TEST_PACKAGE.copy(sourceArtifact = sourceArtifact)
+            val tools = listOf(toolUrl(expUrl, "scancode", SCANCODE_VERSION))
+            stubHarvestTools(wiremock, expUrl, tools)
+            stubHarvestToolResponse(wiremock, expUrl)
+
+            val storage = ClearlyDefinedStorage(storageConfiguration(wiremock))
+            assertValidResult(storage.read(pkg, SCANNER_DETAILS))
+        }
+
+        "deal with package coordinates not supported by ClearlyDefined" {
+            val id = TEST_IDENTIFIER.copy(type = "unknown")
+
+            val storage = ClearlyDefinedStorage(storageConfiguration(wiremock))
+            assertEmptyResult(storage.read(id), expId = id)
+        }
+
+        "handle an unexpected result for the harvest tool request" {
+            stubFor(
+                get(anyUrl())
+                    .willReturn(
+                        aResponse().withStatus(200)
+                            .withBody("This is not a JSON response")
+                    )
+            )
+
+            val storage = ClearlyDefinedStorage(storageConfiguration(wiremock))
+            when (val result = storage.read(TEST_IDENTIFIER)) {
+                is Failure ->
+                    result.error shouldContain "JsonParseException"
+                else -> fail("Unexpected result: $result")
+            }
+        }
+
+        "handle an unexpected result for the harvest tool file request" {
+            val scanCodeUrl = toolUrl(PACKAGE_URL, "scancode", SCANCODE_VERSION)
+            stubHarvestTools(wiremock, PACKAGE_URL, listOf(scanCodeUrl))
+            stubFor(
+                get(urlPathEqualTo("/harvest/$scanCodeUrl"))
+                    .willReturn(
+                        aResponse().withStatus(200)
+                            .withBody("{ \"unexpected\": true }")
+                    )
+            )
+
+            val storage = ClearlyDefinedStorage(storageConfiguration(wiremock))
+            assertEmptyResult(storage.read(TEST_IDENTIFIER))
+        }
+
+        "handle a failure to connect to the server" {
+            val port = ServerSocket(0).use { it.localPort }
+            val serverUrl = "http://localhost:$port"
+
+            val storage = ClearlyDefinedStorage(ClearlyDefinedStorageConfiguration((serverUrl)))
+            when (val result = storage.read(TEST_IDENTIFIER)) {
+                is Failure ->
+                    result.error shouldContain "ConnectException"
+                else -> fail("Unexpected result: $result")
+            }
+        }
+    }
+}) {
+    companion object {
```

Comment:
We should probably consider making all `companion object` members top-level members in the file instead to reduce nesting for better readability.